[Pitch] Strongly Typed Regex Captures v2 (#42)
apple/swift-experimental-string-processing: Documentation/Evolution/StronglyTypedCaptures.md

}
```

>_**Note:** The `Regex` type includes, and `firstMatch(of:)` returns, the entire

> match as the "0th capture"._

Reviewer: In the rest of the document, by "captures" we refer to second tuple element and onwards. For instance:
```swift
(EntireMatch, Capture...)
```
So it's probably better not to refer to the entire match as the "0th capture".

but handling that will likely introduce constraints on `Captures`. We use an
unconstrained generic parameter in this pitch for brevity and simplicity. The
`Substring`s we use for illustration throughout this pitch are created
on-the-fly; the actual memory representation uses `Range<String.Index>`. In this
sense, the `Captures` generic type is just an encoding of the arity and kind of captured content.*

### `firstMatch(of:)` method

The `firstMatch(of:)` method returns a `Substring` of the first match of the
provided regex in the string, or `nil` if there are no matches. If the provided
regex contains captures, the result is a tuple of the matching string and any
captures (described more below).

```swift
extension String {
    public func firstMatch<R: RegexProtocol>(of regex: R) -> R.Captures?
}
```

This signature is consistent with the traditional numbering of backreferences to
capture groups starting at `\1`. Many regex libraries make the entire match
available as the "0th capture". We propose to do the same in order to align the
tuple index numbering with the regex backreference numbering:

```swift

> let scalarRangePattern = /([0-9a-fA-F]+)(?:\.\.([0-9a-fA-F]+))?/
> // Positions in result: 0 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> //                      1 ^~~~~~~~~~~     2 ^~~~~~~~~~~

Reviewer 2: Nit:

```suggestion
let scalarRangePattern = /([0-9a-fA-F]+)(?:\.\.([0-9a-fA-F]+))?/
// Positions in result: 0 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//                      1 ^~~~~~~~~~~~~~     2 ^~~~~~~~~~~~~~
```

By contrast, the proposed Swift version captures all four sub-matches:

```swift
let pattern = /(?:([0-9a-fA-F]+)-?)+/
if let match = "1234-5678-9abc-def0".firstMatch(of: pattern) {
    print(match.1)
}
// Prints ["1234", "5678", "9abc", "def0"]
```

We believe that the proposed capture behavior leads to better consistency with
the meaning of these quantifiers. However, the alternative behavior does have
the advantage of a smaller memory footprint because the matching algorithm would
not need to allocate storage for capturing anything but the last match. As a
future direction, we could introduce some way of opting into this behavior.

#### Alternation: `a|b`

Alternations are used to match one of multiple patterns. An alternation wraps
its underlying pattern's capture type in an `Optional`.

```swift

> let numberAlternationRegex = /([01]+)|[0-9]+|([0-9a-fA-F]+)/
> // Positions in result:     1 ^~~~~~~      2 ^~~~~~~~~~~~~~
> // => `Regex<(Substring, Substring?, Substring?)>`

Reviewer: ```suggestion
let numberAlternationRegex = /([01]+)|[0-9]+|([0-9a-fA-F]+)/
// Positions in result:     0 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//                          1 ^~~~~~~      2 ^~~~~~~~~~~~~~
// => `Regex<(Substring, Substring?, Substring?)>`
```

// Equivalent result builder syntax:
//     let numberAlternationRegex = Pattern {
//         OneOf {
//             OneOrMore(.binaryDigit).capture()
//             OneOrMore(.decimalDigit)
//             OneOrMore(.hexDigit).capture()
//         }
//     }
//     .flattened()

let scalarRangeAlternation = /([0-9a-fA-F]+)\.\.([0-9a-fA-F]+)|([0-9a-fA-F]+)/
// Positions in result:     1 ^~~~~~~~~~~~~~  2 ^~~~~~~~~~~~~~
//                                                           3 ^~~~~~~~~~~~~~
// => `Regex<(Substring, Substring?, Substring?, Substring?)>

// Equivalent result builder syntax:
//     let scalarRangeAlternation = Pattern {
//         OneOf {
//             Group {
//                 OneOrMore(.hexDigit).capture()
//                 ".."
//                 OneOrMore(.hexDigit).capture()
//             }
//             OneOrMore(.hexDigit).capture()
//         }
//     }
//     .flattened()

let nestedScalarRangeAlternation = /(([0-9a-fA-F]+)\.\.([0-9a-fA-F]+))|([0-9a-fA-F]+)/

> // Positions in result:           1 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~   

Reviewer: It would be clearer if we add a line about `0`-th match because down below we show its generic type.  Same suggestions for other occurrences.

By contrast, the proposed Swift version captures all four sub-matches:

```swift
let pattern = /(?:([0-9a-fA-F]+)-?)+/
if let match = "1234-5678-9abc-def0".firstMatch(of: pattern) {
    print(match.1)
}
// Prints ["1234", "5678", "9abc", "def0"]
```

We believe that the proposed capture behavior leads to better consistency with
the meaning of these quantifiers. However, the alternative behavior does have
the advantage of a smaller memory footprint because the matching algorithm would
not need to allocate storage for capturing anything but the last match. As a
future direction, we could introduce a variant of quantifiers for this behavior
that the programmer would opt in for memory-critical use cases.

#### Alternation: `a|b`

> A quantifier wraps its underlying pattern's capture type in either an `Optional`
> or `Array`.
> 

Reviewer 2: ```suggestion
```

```swift
let regex = /([a-z])+/

> // => `Regex<Substring, CaptureCollection<Substring>>`

Reviewer: ```suggestion
// => `Regex<(Substring, CaptureCollection<Substring>)>`
```

//         func firstMatch<R: RegexProtocol, C1, C2>(of regex: R)
//             -> (Substring, C1, C2)? where R.Captures == (C1, C2)
//         ...
//     }
```

For simple regular expressions this had the benefit of aligning the generic
signature more obviously with the captures in the regex.

```swift
let regex = /ab(cd*)(ef)gh/
// => `Regex<(Substring, Substring)>`
```

However, it came with a number of (not necessarily insurmountable) open
questions:

- Will variadic generic tuple splatting preserve element labels?
- Will variadic generic tuple splatting eliminate `Void`s? _(We don't want
`firstMatch(of:)` to return `(Substring, Void)` for a regex with no captures)._
- Will we be able to add [single-element labeled tuples](https://forums.swift.org/t/single-element-labeled-tuples/9797)?
_(This would be needed to preserve the name of a capture in a regex with a
single named capturing group.)_
- What should be the type of `Captures` for a regex with no captures (e.g.
`Void` or `Never` or something else)?

Also, when using a regex within the result builder syntax, it's more obvious
that `capture()` will have access to the entire match, when the match is

> included in the generic signature.

Reviewer: > it's more obvious that `capture()` will have access to the entire match

This point is not so obvious to me. In either the old or the new design, a regex with the result builder syntax can be captured to produce the its entire match in the pattern context.

```swift
let pattern = Pattern {
    /abc/.capture() // either design can capture the entire match of this regex.
}
```

Is this what you are referring to, or something else?

Author: I was thinking about a situation where the regex includes captures:

```swift
let pattern = Pattern {
   /([0-9A-Z]+)(?:\.\.([0-9A-Z]))?/.capture { (match, lower, upper) in ... }
}
```

If we think this point is valuable, then we probably need to include an example to clarify.

Reviewer: Hmmm in this case it still follows that `capture()` will include the entire match as a `Substring` in both designs. That is, if you capture anything it gets included in the generic type in both approaches.

Given all of this, it seems simpler and more pragmatic to make `Regex` generic
over both the match and the captures.

### Structured rather than flat captures

This pitch proposes infering capture types in such a way as to align with the
traditional numbering of backreferences. This is because much of the motivation
behind providing regex literals in Swift is their familiarity.

If we decided to deprioritize this motivation, there are opportunities to infer
safer, more ergonomic, and arguably more intuitive types for captures.

> For example, to be consistent with traditional regex backreferences
> quantifications of multiple or nested captures produces parallel arrays rather
> than an array of tuples.

Reviewer: ```suggestion
For example, to be consistent with traditional regex backreferences
quantifications of multiple or nested captures had to produce parallel arrays
rather than an array of tuples.
```

Given all of this, it seems simpler and more pragmatic to make `Regex` generic
over both the match and the captures.

### Structured rather than flat captures

> This pitch proposes infering capture types in such a way as to align with the

Reviewer: ```suggestion
This pitch proposes inferring capture types in such a way as to align with the
```

traditional numbering of backreferences. This is because much of the motivation
behind providing regex literals in Swift is their familiarity.

If we decided to deprioritize this motivation, there are opportunities to infer
safer, more ergonomic, and arguably more intuitive types for captures.

For example, to be consistent with traditional regex backreferences
quantifications of multiple or nested captures produces parallel arrays rather
than an array of tuples.

```swift
/(?:(?<lower>[0-9a-fA-F]+)\.\.(?<upper>[0-9a-fA-F]+))+/
// Flat captures type:
// => `Regex<(Substring, lower: [Substring], upper: [Substring])>`

// Structured captures type:
// => `Regex<(Substring, [(lower: Substring, upper: Substring)])>`
```

> The structured captures type is safer because the type system encodes that there

Reviewer: There are occurrences of both "captures type" and "capture type". Let's standardize on "capture type"?

To support dynamism, we could introduce a new type, `DynamicCaptures` that
represents a tree of captures, and add a `Regex` initializer that accepts a

> string and produces `Regex<Substring, DynamicCaptures>`.

Reviewer: ```suggestion
string and produces `Regex<(Substring, DynamicCaptures)>`.
```
